core2/pkg/accounting/accounting_internal.go

```diff
@@ -390,6 +391,125 @@ func internalCommitAllocation(b *internalBucket, allocId accAllocId) {
+	}
+	if iParent.OwnedBy != iOwnerByActor.Id {
+		return util.HttpErr(http.StatusForbidden, "You are not allowed to modify this allocation")
+	}
+
+	if iAlloc.Retired {
+		return util.HttpErr(http.StatusForbidden, "You cannot update a retired allocation, it has already expired!")
+	}
+	if iAlloc.Start.Before(now) && newStart.Present {
+		return util.HttpErr(http.StatusForbidden, "You cannot change the starting time of an allocation which has already started")
+	}
+	proposedNewStart := iAlloc.Start
+	if newStart.Present {
+		proposedNewStart = newStart.Value.Time()
+	}
+	proposedNewEnd := iAlloc.End
+	if newEnd.Present {
+		proposedNewEnd = newEnd.Value.Time()
+	}
+	if proposedNewStart.After(proposedNewEnd) {
+		return util.HttpErr(http.StatusForbidden, "This update would make the allocation invalid. An allocation cannot start after it has ended.")
+	}
+	if newQuota.Present && newQuota.Value < 0 {
+		errorMessage := fmt.Sprintf("You cannot set a negative quota for an allocation (%d)", newQuota.Value)
+		return util.HttpErr(http.StatusForbidden, errorMessage)
+	}
+
+	proposedNewQuota := iAlloc.Quota
+	if newQuota.Present {
+		proposedNewQuota = newQuota.Value
+		delta := newQuota.Value - iAlloc.Quota
+		activeQuota := lInternalGroupTotalQuotaFromActiveAllocations(b, iAllocGroup)
+		activeUsage := iAllocGroup.TreeUsage
+
+		if activeQuota+delta < activeUsage {
+			return util.HttpErr(http.StatusForbidden, "You cannot decrease the quota below the current usage!")
+		}
+	}
+
+	iAlloc.Dirty = true
+	iAlloc.Start = proposedNewStart
+	iAlloc.End = proposedNewEnd
+	iAlloc.Quota = proposedNewQuota
+
+	lInternalMarkSignificantUpdate(b, now, iWallet)
+	lInternalReevaluate(b, now, iWallet, true)
+
+	if iAlloc.GrantedIn.Present {
+		comment := ""
+		if newQuota.Present {
+			amount := int64(0)
+			// Converting to core hours in case of minutes or day
+			switch b.Category.AccountingFrequency {
+			case accapi.AccountingFrequencyOnce:
+				amount = proposedNewQuota
+			case accapi.AccountingFrequencyPeriodicMinute:
+				amount = proposedNewQuota / 60
+			case accapi.AccountingFrequencyPeriodicHour:
+				amount = proposedNewQuota
+			case accapi.AccountingFrequencyPeriodicDay:
+				amount = proposedNewQuota * 24
+			default:
+				log.Warn("Invalid accounting frequency passed: '%v'\n", b.Category.AccountingFrequency)
+			}
+			comment += fmt.Sprint("The Quota for %s (%s) has manually been updated to %s.\n", b.Category.Name, b.Category.Provider, strconv.FormatInt(amount, 10))
+		}
+		if newStart.Present {
+			comment += fmt.Sprint("The start date for the granted %s (%s) allocation has manually been updated to %s .\n", b.Category.Name, b.Category.Provider, proposedNewStart.String())
+		}
+		if newEnd.Present {
+			comment += fmt.Sprint("The end date for the granted %s (%s) allocation has manually been updated to %s .\n", b.Category.Name, b.Category.Provider, proposedNewEnd.String())
+		}
+
+		_, err := GrantsPostComment(
```

Comment:
This call shouldn't be made while holding the bucket lock since it will keep it for a very long time. `GrantsPostComment` will immediately call the database and can thus delay all other accounting calls in that bucket for several milliseconds.